## 七月在线机器学习实训营三期第六周(机器学习)考试
#### 考试说明:
- 起止时间：请同学在2018年3月16日至3月18日期间完成，最晚提交时间本周日（3月18日24时之前）结束，<b>逾期不接受补考,该考试分数计入平时成绩</b>
- 考试方式：请同学<font color=red><b>拷贝</b></font>该试卷至自己姓名的目录后，将文件更名为同学姓名拼音-exam6后，进行作答。例如wangwei-exam6
- 提交格式：请同学新建自己姓名全拼的文件夹，将该试卷，数据文件，zip文件等相关考试文件，放置此目录下。将该目录<b>移动</b>至/0.Teacher/Exam/4/目录下
- 注意事项：为确保同学们真正了解自身对本周课程的掌握程度，<font color=red><b>请勿翻阅，移动，更改</b></font>其它同学试卷。如发现按0分处理
- 请同学在下方同学姓名处填写自己的姓名，批改人和最终得分不用填写

- 同学姓名:<u>_邓嘉____</u>  
- 批改人： David
- 最终得分:98

<center><h1>####答卷开始####</h1></center>
<hr>

## 简答题(共4题，每题5分，共计20分)
- note:20

#### 1. 请写出你了解的机器学习特征工程操作，以及它的意义（面试题）

我了解的特征工程是对特征的选择以及特征的前期处理，特征的前期处理包括离散话，独热向量编码等，特征选择的操作比如说过滤式，包裹式，嵌入式以及正则化，其意义1属性过多可能会带来维数灾难问题，若能从中找出重要的特征，维度灾难问题可以大为减轻，2去除不相关的特征往往会降低学习的难度。另外当特征数过少的的时候可能会出现欠拟合状态，可通过增加特征的方式，来解决

#### 2.请写出上述特征工程操作的sklearn或者pandas实现方式


特征数据处理：
sklearn OneHotEncoder
特征选择
过滤式
SelectKBest
包裹式
RFE
嵌入式
SelectFromModel


#### 3.模型评估中的留一法，留出法，交叉验证分别是什么操作？

留一法：
例如样本一共有n个，将样本中取一个作为验证，其余n-1用于训练，这样轮训n次，也就是每个样本都做一次验证集，然后再取这n次的平均值
留出法：
将样本划分为互斥训练和测试两部分样本，用训练集数据训练出来的模型在验证机上进行误差测量
交叉验证：
将样本划分为k个互斥的子集，一个子集用于验证，k－1个子集用户训练，这样就可以获得k组训练集和样本集，最后这k组的平均值


#### 4.如何理解模型的过拟合与欠拟合，以及如何解决？（面试题）

过拟合是指对训练样本集拟合的过于完美，从而失去了泛化能力，对新样本测试不准的问题，欠拟合就是指在训练集上的误差就较大，对数据的拟合不好
过拟合的情况
1 清洗数据
2 增大训练数据
3 采用正则化，正则话包括l1和l2 ，l1均有消除特征的作用，l2减小特征作用的能力

欠拟合的情况：
1 增加其他特征
2 增加多项式特征
3 减小正则话参数

## 操作题(共1题，共计80分)

- note: 78

### 信用卡欺诈项目(共7项，前5项每题10分，6，7题每题15分)

- 前期数据导入,预览及处理(此部分勿修改，涉及的数据文件无需复制移动)

In [70]:
import pandas as pd
pd.set_option('display.max_columns', 500)
import zipfile
with zipfile.ZipFile('./KaggleCredit2.csv.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
    data = pd.read_csv(f, index_col=0)
data.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0                 1                              0.766127  45.0   
1                 0                              0.957151  40.0   
2                 0                              0.658180  38.0   
3                 0                              0.233810  30.0   
4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [71]:
data.shape

(112915, 11)

In [72]:
data.isnull().sum(axis=0)

SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                     4267
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                              0
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                      4267
dtype: int64

In [73]:
data.dropna(inplace=True)
data.shapey = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)

In [74]:
y = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)
y.mean()

0.06742876076872101

- 1.把数据切分成训练集和测试集

In [75]:
from distutils.version import LooseVersion as Version
from sklearn import __version__ as sklearn_version
if Version(sklearn_version) < '0.18':
    from sklearn.cross_validation import train_test_split
else:
    from sklearn.model_selection import train_test_split


In [76]:
X, y = data.iloc[:, 1:].values, data.iloc[:, 0].values

X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.3, random_state=0)

- 2.使用logistic regression建模，并且输出一下系数，分析重要度。   
- note: 终于有人分析了。

In [77]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(C=1000, random_state=0)
lr.fit(X_train,y_train)

LogisticRegression(C=1000, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [78]:
lr.coef_

array([[ -7.13523919e-05,  -4.74145617e-02,   1.27273015e-01,
          2.02464081e-02,  -4.64273874e-05,  -1.00934694e-02,
          3.29487272e-02,   1.26957644e-02,  -1.23859193e-01,
          5.16305419e-02]])

DebtRatio 这一特征比较重要
RevolvingUtilizationOfUnsecuredLines  这一特征影响不大

- 3.使用决策树/SVM/KNN...等sklearn分类算法进行分类，尝试了解参数含义，调整不同的参数。

In [16]:
from sklearn.svm import SVC
svm = SVC(kernel='linear',max_iter=100,C=100.0, random_state=0)
sv=svm.fit(X_train, y_train)


/Users/dengjia1778/Documents/Anaconda/anaconda3/lib/python3.6/site-packages/sklearn/svm/base.py:218: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  % self.max_iter, ConvergenceWarning)


In [17]:
sv.coef_

array([[ -3.85679640e-03,  -1.05279528e-02,  -2.98847070e-02,
         -5.03015294e-02,  -6.15698394e-05,  -7.67985232e-02,
          1.64304142e-02,  -3.57933981e-02,  -3.76162169e-02,
         -3.04598930e-01]])

In [18]:
from sklearn.tree import DecisionTreeClassifier
tree = DecisionTreeClassifier(criterion='entropy', max_depth=3, random_state=0)
tree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=3,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best')

In [20]:
tree.n_features_ 

10

- note: ？

In [21]:
tree.tree_

In [23]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5, p=2, metric='minkowski')
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

- 4.在测试集上进行预测，计算准确度

In [79]:
y_pred=lr.predict(X_test)

In [80]:
lr.score(X_test,y_test)

0.93345605154164746

- 5.查看sklearn的官方说明，了解混淆矩阵等评估标准，并对此例进行评估。

In [81]:
type(y_test)

numpy.ndarray

In [82]:
from sklearn.metrics import confusion_matrix
confus_matrix=confusion_matrix(y_test, y_pred)
confus_matrix

array([[30406,    18],
       [ 2151,    20]])

准确率

In [83]:
auc=(confus_matrix[0][0]+confus_matrix[1][1])/confus_matrix.sum()
auc

0.93345605154164746

误分类率

In [84]:
error_rate=1-auc
error_rate

0.066543948458352542

命中率

In [85]:
ppv=confus_matrix[0][0]/(confus_matrix[0][0]+confus_matrix[0][1])
ppv

0.99940836181961612

负例命中率

In [86]:
npv=confus_matrix[1][1]/(confus_matrix[1][0]+confus_matrix[1][1])
npv

0.0092123445416858584

覆盖率

In [87]:
trr=confus_matrix[0][0]/(confus_matrix[0][0]+confus_matrix[1][0])
trr

0.93393125902263718

负例覆盖率

In [88]:
spc=confus_matrix[1][1]/(confus_matrix[0][1]+confus_matrix[1][1])
spc

0.52631578947368418

从整体来看，准确率达到了0.9335480901978831，很不错，但是，负例命中率 npv=0.023，模型中所有预测是信用卡欺诈的准确率仅为0.023 负例覆盖率 spc=0.526   也就是测试集中所有行用卡欺诈只有0.526被测试出来，所以模型还应该进一步优化

- 6.银行通常会有更严格的要求，因为fraud带来的后果通常比较严重，一般我们会调整模型的标准。   
比如在logistic regression当中，一般我们的概率判定边界为0.5，但是我们可以把阈值设定低一些，来提高模型的“敏感度”   
试试看把阈值设定为0.3，再看看这个时候的混淆矩阵等评估指标。

In [96]:
from sklearn.metrics import classification_report


In [98]:
lr_proba = lr.predict_proba(X_test)
proba = lr_proba[:,1] > 0.3
print("thredhold = 0.3 LR:\n",classification_report(y_test, proba))

thredhold = 0.3 LR:
              precision    recall  f1-score   support

          0       0.93      1.00      0.97     30424
          1       0.54      0.01      0.02      2171

avg / total       0.91      0.93      0.90     32595



- 7.尝试对不同特征的重要度进行排序，通过特征选择的方式，对特征进行筛选。并重新建模，观察此时的模型准确率等评估指标。

In [115]:
from sklearn.feature_selection import SelectKBest
import numpy as np
selectBest = SelectKBest(k=7).fit(X_train,y_train)
X_new=selectBest.transform(X_train)
# transform training set:
print('特征得分：',selectBest.scores_)
print('筛选前：',X_train.shape)
print('筛选后：',X_new.shape)
print(X_new.shape)
np.argsort(selectBest.scores_)

特征得分： [  3.45601489e-01   9.06016697e+02   1.22271967e+03   2.49247531e+02
   2.76076765e+01   9.98085590e+01   1.03282754e+03   5.71396954e+01
   7.40189780e+02   1.76728993e+02]
筛选前： (76053, 10)
筛选后： (76053, 7)
(76053, 7)


array([0, 4, 7, 5, 9, 3, 8, 1, 6, 2])

In [ ]:
X_select=X_test[:,[0,4,7,5,9,3,8,1]]

In [ ]:
lr_probas = lr.predict_proba(X_select)
proba = lr_probas[:,1] > 0.3
print("thredhold = 0.3 LR:\n",classification_report(y_test, proba))

### 本阶段课程意见反馈题(5分送分项，非必答)
#### 请同学围绕以下两点进行回答：
- 自身总结：请您对您自己在本周课程的学习，收获，技能掌握等方面进行一次总结 ，也包括有哪些方面存在哪些不足，欠缺，困惑。作为将来回顾学习路径时的依据。


感觉本周的课程压力还是比较大的，课程的进度很快，各个算法的原理感觉还是一知半解，还需要进一步总结。

- 课程反馈：请就知识点，进度，难易度，教学方式，考试方式及难易度等方面向我们反馈，督促我们进行更有效的改进。

上周的课程我感觉连续性不太好，特别是上海这边的同学，本来上周课程中小阳老师对决策树svm等机器模型提出了很多问题让我们思考，周末北京的ppt上都是对相关内容的进行深入的讲解，但是上海这边的同学是上的神经网络bp相关的课程，感觉内容的连续上很不好，尤其是小阳老师相关的提问，我门思考了，但是没有解答，感觉效果不是很好，感觉就是嘴里的食物咀嚼到一半然后吐掉，又去吃另外的东西。

<hr>
<center><h1>####答卷结束####</h1></center>